### PR TITLE
Skip configuring non-existing network interfaces in openstack cloud

### PR DESCRIPTION
**Pull request by madhuri-rai07**
cloud-init fails to configure network interfaces running with
OpenStack cloud if any of the network interfaces doesn't exist
on the node. This patch fixes the issue by skipping to configure
those interfaces and continue with only the interfaces that are
found on the node.

Closes-Bug: #1857031

### CODE REVIEW

**Comment by igalic** [comment]
the build failure looks transitory, and not something to de concerned with, when submitting a docs patch ;)

**Comment by igalic** [comment]
however, you'll have to link your launchpad user to your github user, and sign the CLA: https://github.com/canonical/cloud-init/blob/master/HACKING.rst

**Comment by OddBloke** [comment]
Hi @madhuri-rai07, thanks for the pull request! As @igalic mentioned, you will need to sign the CLA before we can accept the contribution. If you have any questions about how to go about doing so, please either ask here (and feel free to @ me) or in #cloud-init on Freenode IRC.

Thanks again!

**Comment by madhuri-rai07** [comment]
@OddBloke @igalic I have signed the agreement. Thanks!

**Comment by OddBloke** [comment]
@madhuri-rai07 Thanks for signing it! There's one last thing we need you to do, so that we can be sure that your Launchpad account and your GitHub account are owned by the same person. Please run `tools/migrate-lp-user-to-github`, and it will open up merge proposals in each system referencing one another.

**Comment by OddBloke** [comment]
(I've opened up https://github.com/canonical/cloud-init/pull/130 to clarify the wording, as the existing doc did not make it at all clear that you needed to do this!)

**Comment by madhuri-rai07** [comment]
@OddBloke Thanks! I get this error "fatal: remote error: Repository '~madhuri-rai07/cloud-init' not found."

**Comment by OddBloke** [comment]
@madhuri-rai07 Aha, that's because the script assumes that you already have a Launchpad repo. I'll fix that, but in the meantime I think you can do: `git push git+ssh://<LAUNCHPAD USER NAME>@git.launchpad.net/~madhuri-rai07/cloud-init` to work-around it, then re-run the script.

Apologies for the inconvenience!

**Comment by OddBloke** [comment]
For reference, https://github.com/canonical/cloud-init/pull/136 should fix this issue.

**Comment by madhuri-rai07** [comment]
Hi @OddBloke , still facing issue http://paste.openstack.org/show/787887/

**Comment by OddBloke** [comment]
@madhuri-rai07 Hmm, that's interesting. What OS/desktop environment are you running this on?

In the meantime, I've manually created the MP that the tool otherwise would have: https://code.launchpad.net/~madhuri-rai07/cloud-init/+git/cloud-init/+merge/377187

You'll still need to create a GitHub pull request from the migrate-lp-to-github branch to master, which would be the last step that the script would perform.

**Comment by madhuri-rai07** [comment]
@OddBloke Hi, I am using Ubuntu 18.04

**Comment by madhuri-rai07** [comment]
@OddBloke Hi, I have created a new pull request https://github.com/canonical/cloud-init/pull/156
Thanks!

File: cloudinit/sources/helpers/openstack.py [hunk 1]
```diff
@@ -674,7 +674,9 @@ def convert_net_json(network_json=None, known_macs=None):
             if not mac:
                 raise ValueError("No mac_address or name entry for %s" % d)
             if mac not in known_macs:
-                raise ValueError("Unable to find a system nic for %s" % d)
+                LOG.error("Unable to find a system nic for %s. Skipping..." % d)
+                config.remove(d)
+                continue
```
Comment (smoser) [inline cloudinit/sources/helpers/openstack.py]: As it is, this is simply not OK. swallowing this error isn't really helping anyone.
The cloud platform told us there was a NIC that needed to be configured, if cloud-init merely logs the error, other things will break later and the user have to search the logs for "what went wrong".

possibly you could add a config option to cloud-init to "skip missing nics" or something, but the default behavior should be to fail.

Comment (madhuri-rai07) [reply cloudinit/sources/helpers/openstack.py]: @smoser that makes sense to me. I will revise the patch to include a config parameter. Do you have any reference patch which adds a config to the cloud-init project?
P.S I am new to cloud-init code :)

**Comment by OddBloke** [comment]
Hi @madhuri-rai07, thanks for your work on this contribution. As this hasn't seen any forward movement in a while, I'm going to close this out. If you're still interested in contributing this change (with the config option, as requested), please do feel free to reopen it and ask any questions that you might have.

**Pull request closed without merging**